src/tokio_reactor.rs

```diff
@@ -0,0 +1,222 @@
+};
+
+use async_trait::async_trait;
+use tokio::sync::mpsc;
+
+use core::time::Duration;
+use std::thread;
+use std::time::SystemTime;
+
+/// Communication bridge with a consensus reactor.
+pub struct TokioReactor {
+    timeout: Duration,
+
+    listener: mpsc::Receiver<Message>,
+    sender: mpsc::Sender<Message>,
+
+    /// Reactor will dispatch messages to
+    outbound: mpsc::Sender<Message>,
+}
+
+impl TokioReactor {
+    /// Await for the next message sent from a reactor
+    pub async fn next_async(&mut self) -> Option<Message> {
+        self.listener.recv().await
+    }
+
+    /// Send a notification to the reactor
+    pub async fn notify(&mut self, notification: Notification) -> Result<(), Error> {
+        let notification = Message::Notification(notification);
+
+        self.sender
+            .send_timeout(notification, self.timeout)
+            .await
+            .map_err(|_| Error::ResourceNotAvailable)
+    }
+
+    /// Send a request to the reactor
+    pub async fn request(&mut self, request: Request) -> Result<Response, Error> {
+        let id = request.id();
+        let request = Message::Request(request);
+
+        let start = SystemTime::now();
+        let mut requeue = Vec::new();
+
+        self.sender
+            .send_timeout(request, self.timeout)
+            .await
+            .map_err(|_| Error::ResourceNotAvailable)?;
+
+        #[cfg(feature = "trace")]
+        tracing::debug!(
+            "request {:?} sent, awaiting response with timeout {:?}",
+            request,
+            self.timeout
+        );
+
+        loop {
+            match start.elapsed() {
+                Err(_e) => {
+                    requeue.into_iter().for_each(|m| {
+                        self.outbound.try_send(m).ok();
+                    });
+
+                    #[cfg(feature = "trace")]
+                    tracing::debug!("request {} failed with error {}", id, _e);
+
+                    return Err(Error::ResourceNotAvailable);
+                }
+
+                Ok(elapsed) if elapsed > self.timeout => {
+                    requeue.into_iter().for_each(|m| {
+                        self.outbound.try_send(m).ok();
+                    });
+
+                    #[cfg(feature = "trace")]
+                    tracing::debug!("request {} failed with timeout", id);
+
+                    return Err(Error::ResourceNotAvailable);
+                }
+
+                _ => (),
+            }
+
+            #[cfg(feature = "trace")]
+            tracing::trace!("attempting to receive request {}", id);
+
+            while let Some(m) = self.listener.try_recv().ok() {
+                match m {
+                    Message::Response(r) if r.id() == id => {
+                        requeue.into_iter().for_each(|m| {
+                            self.outbound.try_send(m).ok();
+                        });
+
+                        return Ok(r);
```

Comment:
@Voxelot Isn't this done line above with `requeue`?

The edge case when I look at this again is if `outbound` is near full, `try_send` can fail while we shufling requeue and wait for our response somebody else can fill it, in this case we would drop those `Messages` that does not fit inside.

Reply:
ah you're right, I missed the above section. We should probably log a warn message here if we get a TrySendError, so we know that if something like this happens we can increase the buffer size. 

Since we have the potential for messages to drop and requestors may never receive a response, we might also want to add a timeout to this function.

Reply:
As discussed, since this is BFT, we are tolerant to eventual failures. We avoid risks of overloading the queue in case we have some accumulated pending messages by retrying only once.

Reply:
You are here "potentially" dropping a response, it is really an edge case, but randomly dropping responses is never a good idea, in general, if the channel queue gets full isn't it on the sender to decide what he wants to do, for example, `fuel-core-p2p` will just drop consensus votes it does not care for it while `fuel-core-sync` would wait for validation.
Not sure how to exactly do that in this case, maybe leave it for some future pull requests?

Reply:
We don't drop messages at random, they are requeued here: https://github.com/FuelLabs/fuel-bft/blob/0fe5b880cb94e78eaa8e67aa141c9049ea79439c/src/tokio_reactor.rs#L91-L93

They will be dropped only if the tokio channel capacity is choked

Reply:
by random, I meant random response if the channel gets at a full capacity. 

tldr: It is better to drop the Request than the Response if the channel gets full, that way the requester will know if the request got through or not.

Reply:
since request is a debug interface the risk of this impacting the system adversely is low.